### ISSUE TITLE
$item->link on mod_k2_content creating links with appropriate itemID, SEF links fail

### ISSUE THREAD

**Issue by kbrookes** [joomlaworks/k2-v3-dev-build#251]
Links for $item->link generated in mod_k2_content look like this:

index.php?option=com_k2&view=item&id=2:item-name&Itemid=

When SEF URLs, a 404 error is generated.

**Comment by kavadas** [comment]
I cannot verify the issue. The link indeed should be like that when there is no matching menu item. However, the link is working fine for me both in SEF and non SEF URLs.

**Comment by kbrookes** [comment]
I should have realised about the itemID - it's not like I haven't had to create hidden menus to create that sort of behaviour in the past - thank you.

But here's where it gets odd. I realise I neglected to add the SEF generated URL to the original post (note that at this stage it's just core Joomla SEF, without .htaccess turned on), here it is:

index.php/component/k2/item/2-item-name

That's where I get the 404.

I created a hidden menu item to the K2 item's parent category.

And the 404 went away and generates this URL: index.php/category-name/item-name

Turned off the menu item: 404 again on both:

index.php/category-name/item-name
index.php/component/k2/item/2-item-name

Note that the category/items are all published.

So at this stage on my development, it's looking like mod_k2_content $item->link needs to have at least a menu item to the category published to be able to click through to the content otherwise a 404 is generated.

**Comment by kavadas** [comment]
I still cannot verify that. The URLs like index.php/component/k2/item/2-item-name are working fine for me. 

Here are my settings, in case we have something different:

Search Engine Friendly URLs -> Yes
Use URL Rewriting -> No
Adds Suffix to URL -> No
Unicode Aliases -> No

In K2 settings the "Enable advanced SEF for K2 URLs" is set to "No".

**Comment by kbrookes** [comment]
It's to do with the K2 advanced SEF URLs.

I have the same system SEF:

https://www.dropbox.com/s/gyls5nkjgwvkzpj/Screenshot%202015-09-22%2018.18.04.png?dl=0

But I have the Advanced SEF for K2 URLs enabled and set to 'normal' mode.

Here are the test cases:

1) Turning off K2 Advanced SEF / publishing hidden menu to category - Link works

2) Turning off K2 Advanced SEF / unpublishing hidden menu to category - Link works

3) Turning on K2 Advanced SEF / unpublishing hidden menu to category - Link fails

4) Turning on K2 Advanced SEF / publishing hidden menu to category - Link works

**Comment by kavadas** [comment]
If you enable K2 advanced SEF you have to create hidden menu links to the "General" view and then use them as prefixes in the settings there. Have you done this?

**Comment by kbrookes** [comment]
No - I can see now how that works.

I should note that it **is** working with a hidden menu item just to the item's parent category. 

This seems to give me a correct SEF category part of the URL

I assume that creating a 'general' menu item and setting a predefined alias in the SEF settings is a 'catch all' for instances where you fail to do this?

**Comment by kavadas** [comment]
Yes, that's how it works. You just define a menu link to match all the items ( or categoies, tags, users ) that do not resolve to any existing menu link. In addition to that, advanced SEF allows you to define the pattern for the resource id and alias. 

In any case if you enable advanced SEF you are required to create the menu links to the general view. This needs to be documented or to provide a more clear tip in the settings.

**Comment by kbrookes** [comment]
Great - that makes sense, thank you.

I think a tip in the SEF settings would be great, although I'm now not likely to forget it :)
